runtime/router.go

```diff
@@ -134,77 +58,60 @@ func NewRouterEndpoint(
 	handlerName string,
 	handler HandlerFn,
 ) *RouterEndpoint {
-	endpointTags := map[string]string{
+	scope := gateway.AllHostScope.Tagged(map[string]string{
 		"endpoint": endpointName,
 		"handler":  handlerName,
-	}
-	endpointScope := gateway.AllHostScope.Tagged(endpointTags)
-	requestRecvd := endpointScope.Counter("inbound.calls.recvd")
-	requestLatency := endpointScope.Timer("inbound.calls.latency")
-	statusCodes := make(map[int]tally.Counter, len(knownStatusCodes))
-
-	for _, statusCode := range knownStatusCodes {
-		metricName := "inbound.calls.status." + strconv.Itoa(statusCode)
-		statusCodes[statusCode] = endpointScope.Counter(metricName)
-	}
-
+	})
 	return &RouterEndpoint{
-		EndpointName: endpointName,
-		HandlerName:  handlerName,
-		HandlerFn:    handler,
-		gateway:      gateway,
-
-		metrics: EndpointMetrics{
-			requestRecvd:   requestRecvd,
-			statusCodes:    statusCodes,
-			requestLatency: requestLatency,
-		},
+		HandlerFn: handler,
+		gateway:   gateway,
+		metrics:   NewInboundHTTPMetrics(scope, gateway.Logger),
 	}
 }
 
 // HandleRequest is called by the router and starts the request
 func (endpoint *RouterEndpoint) HandleRequest(
 	w http.ResponseWriter, r *http.Request, params httprouter.Params,
 ) {
-	reqFields := logRequestFields(r)
-	var resFields []zapcore.Field
-
+	var req *ServerHTTPRequest
 	defer func() {
-		writeLogs(endpoint.gateway.Logger, reqFields, resFields)
+		// Post-process logs & metrics
+		writeLogs(endpoint.gateway.Logger, logRequestFields(r), logResponseFields(req.res))
+		endpoint.metrics.CollectMetrics(req.startTime, req.res.finishTime, req.res.StatusCode)
```

Comment:
let's not defer metrics collection and do it inline in `finish()` and `start()`

Reply:
We have a cleaner separation of concerns by not embedding the metrics collection in the res/req.  The req/res also doesn't need to spend time collecting metrics before it writes a response to the wire.  We write to the wire quickly, then record logs/metrics extra.  The defer delay to emit metrics or logs will be negligible.